### PR TITLE
Add Trait `Ceil` to migration built-in functions like `ceil_*`

### PR DESCRIPTION
**Pull request by iosmanthus**
Signed-off-by: Iosmanthus Teng <myosmanthustree@gmail.com>

###  What have you changed?
Add trait `Ceil` to migrate scalar functions:
* `ceil_real`
* `ceil_dec_to_dec`
* `ceil_dec_to_int`
* `ceil_int_to_int`

###  What is the type of changes?

Pick one of the following and delete the others:
- New feature (a change which adds functionality)

###  How is the PR tested?
unit test suite.

### CODE REVIEW

File: components/tidb_query/src/expr/ctx.rs [hunk 1]
```diff
@@ -244,7 +244,7 @@ impl EvalContext {
         if !is_overflow {
             Ok(())
         } else {
-            self.handle_overflow_err(Error::overflow("DECIMAL", ""))
+            self.handle_overflow_err(Error::overflow("", ""))
```
Comment (sticnarf) [inline components/tidb_query/src/expr/ctx.rs]: It's a problem indeed that `handle_overflow` only works for Decimal but I think we should change the function name or remove the function.

File: components/tidb_query/src/rpn_expr/impl_math.rs [hunk 2]
```diff
@@ -2,8 +2,72 @@ use tidb_query_codegen::rpn_fn;
 
 use crate::codec::data_type::*;
 use crate::codec::{self, Error};
+use crate::expr::EvalContext;
 use crate::Result;
 
+#[rpn_fn(capture = [ctx])]
+pub fn ceil<C: Ceil>(ctx: &mut EvalContext, arg: &Option<C::Input>) -> Result<Option<C::Output>> {
```
Comment (sticnarf) [inline components/tidb_query/src/rpn_expr/impl_math.rs]: It is quite common that a function taking a null always returns null. Can we just abstract this behavior?

Comment (breeswish) [reply components/tidb_query/src/rpn_expr/impl_math.rs]: Maybe @andylokandy can work on this later.

Comment (mapleFU) [reply components/tidb_query/src/rpn_expr/impl_math.rs]: > Maybe @andylokandy can work on this later.

Should we just use a macro like `try_opt`?

File: components/tidb_query/src/rpn_expr/impl_math.rs [hunk 3]
```diff
@@ -112,4 +176,82 @@ mod tests {
...
+                .unwrap();
+            assert_eq!(expected, output);
+        }
+    }
+
+    #[test]
+    fn test_ceil_dec_to_dec() {
+        let cases = vec![
+            ("9223372036854775808", "9223372036854775808"),
+            ("-123", "-123.456"),
```
Comment (mapleFU) [inline components/tidb_query/src/rpn_expr/impl_math.rs]: What about the cases when `input` is `"123.456"`?

**Comment by sticnarf** [comment]
> https://stackoverflow.com/questions/37639276/when-should-inline-be-used-in-rust
> 
> So should we use `#[inline]` for your functions?

I think no matter we add `#[inline]`, the compiler should be smart enough to inline such small functions...I think there is no drawback if we keep them.

File: components/tidb_query/src/expr/ctx.rs [hunk 4]
```diff
@@ -240,7 +240,7 @@ impl EvalContext {
         Err(err)
     }
 
-    pub fn handle_overflow(&mut self, is_overflow: bool) -> Result<()> {
```
Comment (lonng) [inline components/tidb_query/src/expr/ctx.rs]: Seems we can remove this method.

File: components/tidb_query/src/codec/mysql/decimal.rs [hunk 5]
```diff
@@ -83,7 +83,7 @@ impl<T> Res<T> {
             Res::Overflow(t) => if let Some(error) = overflow_err {
                 ctx.handle_overflow_err(error)
             } else {
-                ctx.handle_overflow(true)
+                ctx.handle_overflow_err(Error::overflow("", ""))
```
Comment (lonng) [inline components/tidb_query/src/codec/mysql/decimal.rs]:
```suggestion
                ctx.handle_overflow_err(Error::overflow("DECIMAL", ""))
```

File: components/tidb_query/src/rpn_expr/impl_math.rs [hunk 6]
```diff
@@ -112,4 +182,83 @@ mod tests {
...
+        }
+    }
+
+    #[test]
+    fn test_ceil_dec_to_int() {
+        let cases = vec![
+            (124, "123.456"),
+            (2, "1.23"),
+            (-1, "-1.23"),
+            (std::i64::MIN, "-9223372036854775808"),
```
Comment (sticnarf) [inline components/tidb_query/src/rpn_expr/impl_math.rs]: Ceil overflow and integer overflow case missing

File: components/tidb_query/src/rpn_expr/impl_math.rs [hunk 7]
```diff
@@ -112,4 +182,83 @@ mod tests {
...
+        }
+    }
+
+    #[test]
+    fn test_ceil_dec_to_dec() {
+        let cases = vec![
+            ("9223372036854775808", "9223372036854775808"),
+            ("124", "123.456"),
+            ("-123", "-123.456"),
+        ];
```
Comment (sticnarf) [inline components/tidb_query/src/rpn_expr/impl_math.rs]: Ceil overflow case missing

**Pull request merged**